pkg/ingester/instance_test.go

```diff
@@ -135,6 +137,87 @@ func TestConcurrentPushes(t *testing.T) {
 	// test passes if no goroutine reports error
 }
 
+func TestStreamRates(t *testing.T) {
+	limits, err := validation.NewOverrides(defaultLimitsTestConfig(), nil)
+	require.NoError(t, err)
+	limiter := NewLimiter(limits, NilMetrics, &ringCountMock{count: 1}, 1)
+
+	inst, err := newInstance(defaultConfig(), defaultPeriodConfigs, "test", limiter, loki_runtime.DefaultTenantConfigs(), noopWAL{}, NilMetrics, &OnceSwitch{}, nil)
+	require.Nil(t, err)
+
+	const (
+		concurrent          = 10
+		iterations          = 100
+		entriesPerIteration = 100
+	)
+
+	uniqueLabels := map[string]bool{}
+	startChannel := make(chan struct{})
+	hashes := map[uint64]uint64{}
+
+	var buf []byte
+	wg := sync.WaitGroup{}
+	for i := 0; i < concurrent; i++ {
+		l := makeRandomLabels()
+		for uniqueLabels[l.String()] {
+			l = makeRandomLabels()
+		}
+		uniqueLabels[l.String()] = true
+
+		hashWithoutShard, _ := l.HashWithoutLabels(buf, distributor.ShardLbName)
+		hashes[l.Hash()] = hashWithoutShard
+
+		wg.Add(1)
+		go func(labels string) {
+			defer wg.Done()
+
+			<-startChannel
+
+			tt := time.Now().Add(-5 * time.Minute)
+
+			for i := 0; i < iterations; i++ {
+				err := inst.Push(context.Background(), &logproto.PushRequest{Streams: []logproto.Stream{
+					{Labels: labels, Entries: entries(entriesPerIteration, tt)},
+				}})
+
+				require.NoError(t, err)
+
+				tt = tt.Add(entriesPerIteration * time.Nanosecond)
+			}
+		}(l.String())
+	}
+
+	time.Sleep(100 * time.Millisecond)
```

Comment:
just to see if I understand this: this sleep holds all iteration go routines to make sure they start at a similar time. Is that it?

Reply:
That makes this goroutine sleep until the others are ready It's not strictly necessary